File: pkg/listener/src/process.ts
import fs from "node:fs"
import path from "node:path"

import { SoundStack } from "./soundStack.ts"
import { username, date as parseDate } from "./fileData.ts"
import { isPlayer } from "./roster.ts"
import { loadProcessed, saveProcessed } from "./state.ts"
import { mergeAudio } from "./audio.ts"
import { transcribe } from "./transcribe.ts"
import { run } from "./exec.ts"
import { dataPath, tmpPath, incomingPath, stateFile, pythonDir } from "./paths.ts"
import type { Segment } from "./types.ts"

// Hybrid orchestrator: the TS side of the listener pipeline. Watches Craig zips,
// merges audio (ffmpeg), delegates transcription to python/transcribe.py
// (whisperx), assembles the per-user segments into one ordered script.json, and
// publishes to data/saved/{date}/. Set LISTENER_KEEP_ZIP=1 to preserve source
// zips (handy when validating against real recordings).

function log(msg: string): void {
  console.log(`[listener] ${msg}`)
}

function listZips(dir: string): string[] {
  let names: string[]
  try {
    names = fs.readdirSync(dir)
  } catch {
    log(`incoming dir not readable: ${dir}`)
    return []
  }
  return names.filter((n) => n.endsWith(".zip")).map((n) => path.join(dir, n))
}

function playerTracks(dir: string): string[] {
  return fs
    .readdirSync(dir)
    .filter((n) => n.endsWith(".aac"))
    .filter((n) => isPlayer(username(path.basename(n, ".aac"))))
    .map((n) => path.join(dir, n))
}

function processZip(zipPath: string): boolean {
  const stem = path.basename(zipPath, ".zip")
  const sessionDate = parseDate(stem)
  if (sessionDate === "") {
    log(`skipping ${stem}: could not parse a date from the zip name`)
    return false
  }

  log(`processing ${stem} -> session ${sessionDate}`)
  fs.mkdirSync(tmpPath, { recursive: true })
  const work = fs.mkdtempSync(path.join(tmpPath, "session-"))

  try {
    run("unzip", ["-o", "-q", zipPath, "-d", work])

    const tracks = playerTracks(work)
    if (tracks.length === 0) {
      log(`no player tracks found in ${stem}; skipping`)
      return false
    }
    log(`found ${tracks.length} player track(s)`)

    const outDir = path.join(dataPath, "saved", sessionDate)
    fs.mkdirSync(outDir, { recursive: true })

    log("merging audio (ffmpeg)")
    mergeAudio(tracks, path.join(outDir, "audio.mp3"))

    log("transcribing (whisperx — the slow step)")
    transcribe(pythonDir, outDir, tracks)

    log("assembling script.json")
    const stack = new SoundStack()
    for (const track of tracks) {
      const trackStem = path.basename(track, ".aac")
      const segPath = path.join(outDir, `${trackStem}.json`)
      const segments = JSON.parse(fs.readFileSync(segPath, "utf8")) as Segment[]
      stack.add(username(trackStem), segments)
    }
    fs.writeFileSync(path.join(outDir, "script.json"), JSON.stringify(stack.drain()))

    // Archive the raw player tracks alongside the output (mirrors the original
    // saved/{date}/ layout).
    for (const track of tracks) {
      fs.renameSync(track, path.join(outDir, path.basename(track)))
    }

    log(`done: ${outDir}`)
    return true
  } finally {
    fs.rmSync(work, { recursive: true, force: true })
  }
}

function main(): void {
  log(`incoming=${incomingPath}`)
  log(`data=${dataPath}`)

  const zips = listZips(incomingPath)
  const processed = loadProcessed(stateFile)
  const todo = zips.filter((z) => !processed.has(path.basename(z)))
  log(`found ${zips.length} zip(s), ${todo.length} unprocessed`)

  for (const zip of todo) {
    const name = path.basename(zip)
    try {
      processZip(zip)
      processed.add(name)
      saveProcessed(stateFile, processed)
      if (process.env.LISTENER_KEEP_ZIP !== "1") fs.rmSync(zip, { force: true })
    } catch (err) {
      log(`ERROR processing ${name}: ${err instanceof Error ? err.message : String(err)}`)
    }
  }
}

main()

File: pkg/listener/python/process.py
import json
from zipfile import ZipFile
from loguru import logger as log
from pydub import AudioSegment
import whisperx

from consts import DATA_PATH, DEVICE, INCOMING_PATH, PLAYERS
from db import DB
from file_data import FileData
from file_utils import TempDir, create_data_folder, get_files, move_file, delete_file
from models import get_alignment_model, get_transcription_model
from sound_stack import SoundStack


@log.catch
def main():
    log.info("running craig output processing")

    all_files = get_files(INCOMING_PATH, "zip")
    log.info("found {} files", len(all_files))

    with DB() as db:
        to_process = all_files.difference(db.processed)
        log.info("processing {} files", len(to_process))

        for fd in to_process:
            process_zip(db, fd)


def process_zip(db: DB, fd: FileData):
    log.info("processing zip file {}", fd.file_stem)

    with TempDir() as temp_path:
        log.info("unzipping into {}", temp_path)
        with ZipFile(fd.file_path, "r") as zip_ref:
            zip_ref.extractall(temp_path)
            log.info("complete unzipping file {}", fd.file_stem)

        data_path = f"{DATA_PATH}/saved/{fd.date()}"
        log.info(f"creating path: {data_path}")
        create_data_folder(data_path)

        merge_sound_files(temp_path, data_path)

        log.info("transcribing audio")
        stack = SoundStack()
        for file in get_files(temp_path, "aac"):
            res = transcribe(data_path, file)
            if res is not None:
                username, segments = res
                stack.add(username, segments)

        script = stack.drain()
        with open(f"{data_path}/script.json", "w") as t:
            json.dump(script, t)

        db.add_processed(fd)
        delete_file(fd.file_path)


def merge_sound_files(temp_path: str, data_path: str):
    log.info("merging sound files")

    sounds = []
    longest_ms: int = 0
    for file in get_files(temp_path, "aac"):
        if not any([username in file.file_stem for username in PLAYERS.values()]):
            log.info(f"skipping {file.file_stem}")
            continue

        s: AudioSegment = AudioSegment.from_file(file.file_path)
        longest_ms = max(longest_ms, int(s.duration_seconds) * 1000)
        sounds.append(s)

    log.info("generating audio file ({})", longest_ms / 1000.0)
    audio_file = AudioSegment.silent(duration=longest_ms)
    for sound in sounds:
        audio_file = audio_file.overlay(sound)

    audio_file.export(f"{data_path}/audio.mp3", format="mp3")
    log.info("completed audio file merging")


def transcribe(data_path: str, file: FileData):
    if not any([username in file.file_stem for username in PLAYERS.values()]):
        log.info("skipping {}", file.file_stem)
        return None

    log.info("loading audio file {} ({})", file.file_stem, file.username())
    audio = whisperx.load_audio(file.file_path)

    log.info("transcribing {}", file.file_stem)
    result = get_transcription_model().transcribe(
        audio,
        batch_size=16,
    )

    log.info("aligning {}", file.file_stem)
    alignment_model, alignment_metadata = get_alignment_model()

    result = whisperx.align(
        result["segments"],
        alignment_model,
        alignment_metadata,
        audio,
        DEVICE,
        return_char_alignments=False,
    )

    with open(f"{data_path}/{file.username()}~{file.index()}.json", "w") as t:
        json.dump(result["segments"], t)
    
    move_file(file.file_path, f"{data_path}/{file.file_name}")

    log.info("completed transcribing {}", file.username())
    return (file.username(), result["segments"])


if __name__ == "__main__":
    main()
